internal/wasm/memory.go

```diff
@@ -56,19 +57,63 @@ type MemoryInstance struct {
 	// waiters implements atomic wait and notify. It is implemented similarly to golang.org/x/sync/semaphore,
 	// with a fixed weight of 1 and no spurious notifications.
 	waiters sync.Map
+
+	mmappedMemory bool
+	closed        bool
 }
 
 // NewMemoryInstance creates a new instance based on the parameters in the SectionIDMemory.
-func NewMemoryInstance(memSec *Memory) *MemoryInstance {
+func NewMemoryInstance(memSec *Memory, compilerEngine bool) *MemoryInstance {
 	min := MemoryPagesToBytesNum(memSec.Min)
 	capacity := MemoryPagesToBytesNum(memSec.Cap)
+
+	var buffer []byte
+	var cap uint32
+	mmappedMemory := false
+	if memSec.IsShared && compilerEngine {
+		// When using compiler, memory accesses can happen at the same time that memory is resized, meaning
+		// we cannot have the memory base move during operation. mmap allows allocating memory virtually so
+		// we can grow without changing the base. The spec requires max for shared memory currently because
+		// all threads implementations are effectively expected to use mmap for shared memory.
+		max := MemoryPagesToBytesNum(memSec.Max)
+		b, err := platform.MmapMemory(int(max))
+		if err != nil {
+			panic(fmt.Errorf("unable to mmap memory: %w", err))
+		}
+		sp := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+		sp.Len = int(MemoryPagesToBytesNum(memSec.Min))
+		buffer = b
```

Comment:
Can't we just do `buffer = b[:MemoryPagesToBytesNum(memSec.Min)]`, instead of using `sp`?